### ISSUE TITLE
Conflict With CSSNext

### ISSUE THREAD

**Issue by trisys3** [shellscape/postcss-less#78]
* Node Version: 4
* NPM Version: 3.10.10
* postcss-less Version: 0.16.0 (also tried 1.0.1, which is not officially compatible with Node 4)

If you have a large amount of code to share which demonstrates the problem you're experiencing, please provide a link to your
repository rather than pasting code. Otherwise, please paste relevant short snippets below.

### LESS

```less
// less that demonstrates the issue
```

/******** PASSING ********/
```
.mixin() {
    .class& {
        background-color: black;
    }
}

body {
    .mixin();
}
```

/******** FAILING ********/
```
.mixin() {
    // this line fails
    &.class {
        background-color: black;
    }
}

body {
    .mixin();
}
```

### JavaScript

```js
// js that reproduces the issue
```

/** gulpfile.js **/

```
const gulp = require('gulp');
const lessParser = require('postcss-less');
const next = require('postcss-cssnext');
const postcss = require('gulp-postcss');

const plugins = [
    next(),
];

const options = {
    syntax: lessParser,
};

gulp.task('less', () =>
    gulp.src('app.less')
        .pipe(postcss(plugins, options))
        .pipe(gulp.dest('dest/')));
```

In the console, run `gulp less`.

postcss-cssnext version: 2.11.0
gulp-postcss version: 6.4.0

```
// actual error output, if error was thrown
```

```
Potentially unhandled rejection [4] Unrecognised input in file /repo/folder/app.min.less line no. 122
```

### Expected Behavior

Builds with no errors.

### Actual Behavior

Builds with a single error (see error).

### How can we reproduce the behavior?

Install all dependencies, run gulp with the gulpfile provided and the LESS provided.

NOTE: This is not an issue when using [postcss-scss](https://github.com/postcss/postcss-scss).

I know the intent of this plugin is not to be a LESS compiler. I had to close one of my [other issues](https://github.com/ben-eb/cssnano/issues/373) because I did not fully understand that. That being said, I think `cssnext` is a case where parsing LESS applies, as it is essentially a transpiler. Transpiling does not make sense to do after preprocessing with something like LESS or SCSS, and in fact most of the new CSS features outlined in `cssnext` will fail if you try to put them into LESS. If I am wrong on any of this, and this does not fit the plugin's purpose or you do not care, feel free to tell me.

I am still not sure whether this issue belongs here or on the [cssnext](https://github.com/MoOx/postcss-cssnext) repository, as the error only shows up if I combine both this plugin and `cssnext`. To see for yourself, try taking out `cssnext` and building again.

**Comment by shellscape** [comment]
Is there a stack available to identify which component is throwing the error?

**Comment by trisys3** [comment]
No, but I am pretty sure it is either `postcss-less` or `cssnext`. Maybe there is some sort of better error handler for `postcss` or `gulp`?

**Comment by shellscape** [comment]
I've no idea, but with a message like "Potentially unhandled rejection [4]," try and track down the culprit of that first and it might lead you to the actual error.

**Comment by trisys3** [comment]
I am not sure I can help with the error handling, all I ever see from `less` is the syntax error and the line in the source LESS, which may not even be the original error. However, the error comes when you use an ampersand as the first character of a selector inside a mixin. This works when compiling with the actual LESS compiler, which is why I was surprised. I will try to strip down the original example a bit more, to illustrate.

On the error-handler note, do you know if there actually is a way to show the full error stack when compiling LESS (or rather trying and failing)?

**Comment by shellscape** [comment]
Well I'm not in the business of debugging other folks' modules for postcss-less when other projects exist for that purpose; postcss-less is nuanced and and difficult as it is :) I'm not sure what the reasoning behind using both cssnext _and_ postcss-less is, but that seems mighty edge-casey on the surface. Folks typically use one or the other.

You could try running the css through cssnext first, and then postcss-less separately, outside of gulp (the docs show you how) as a start.

**Comment by trisys3** [comment]
To be honest, it might be an edge case. Essentially, the only reason we originally used LESS was because Bootstrap used LESS. Otherwise, I would have advocated going with Stylus. Bootstrap also has an SCSS port, but it was decided that was not stable enough, or maybe because it used Ruby mainly. Either way, we decided not to use it.

About `cssnext`, you are right, it does not pair very well with LESS. Until now, I have only been able to use `:not`, though with transpilation before the LESS compilation, maybe that can change. The way I see it, the only things LESS provides are variables, mixins, and some functions, all of which are also provided by CSS4, and therefore `cssnext`. However, Bootstrap provides its mixins and variables in LESS (obviously), and we still want to use those. They were saying Bootstrap might move to PostCSS in version 5, but with version 4 taking over a year by now, and not even in beta yet, I do not know how long that will take. There is a Bootstrap port in PostCSS, but we can't really use it until it becomes more stable, if that ever happens.

So basically, I am using `postcss-less` as a stopgap measure until we can move to SCSS for Bootstrap 4. Even then, we have so many LESS variables/mixins it could take months to move over. Until then, I would like to use PostCSS before the LESS compilation, which right now is only `flexbugs` and `cssnext`.

**Comment by shellscape** [comment]
It might be worthwhile then to target less files and cssnext files independently. Probably going to end up being a more complex gulp file with two passes of postcss, but I'd wager you'd avoid this unknown. Still, if you manage to track down the cause and it's something in postcss-less, I'll happily take a look at a stopgap for you

**Comment by trisys3** [comment]
It looks like the error comes from the LESS compilation itself, which I will put into the question. I will also put some simple output, using `gulp-tap`. It looks like what is happening is `cssnext` thinks the mixin is an actual selector (maybe it allows parentheses in selectors?). When it sees the ampersand, `cssnext` adds a second selector. LESS is OK with the first selector, without the extra class, which it sees as a mixin, but it does not like the 2nd selector.

I think all the plugins are doing the right thing, so I am going to close this issue, but I will still update the question to better highlight the error. If anything, the error, if there is one, is in `cssnext`, so this issue really belongs there.

**Comment by shellscape** [comment]
Cool. Thanks for working through this one with me.

**Comment by trisys3** [comment]
No problem. Sorry I couldn't be of more help and that the issue is not really with you. Great plugin, though, and you did a great job updating it. When you started maintaining it and it wasn't marked "experimental" anymore, I jumped at the chance to use it, and it looks like my trust was not misplaced. Not that I wouldn't have used it before, but in the web business, you never know what could happen with an outdated module.

**Comment by trisys3** [comment]
It looks like passing `nesting: false` to the `cssnext` "features" option does the trick. LESS already nests, so we don't need `cssnext` to nest, too. I would suggest a note in the docs, but you were right about this being an edge case, and I am not sure what the note should say, anyway.

Thanks for bearing with me on this.
